### PR TITLE
refactor(storage): separate writing from builder.

### LINKED ISSUE #493: storage: compactor panics when no rows are in new RowSet
Mentioned in #489 and #492 , the `disk::_blob_slt_` panics occasionally. I tried to reproduce the bug on 376e05b73bb865798e8b43af5e55e67c4f994431 .

It's not steadily reproducible. I tried several times and finally caught the stacktrace.

```
* thread #46, name = 'tokio-runtime-worker', stop reason = breakpoint 1.1
  * frame #0: 0x0000000100aaed24 sqllogictest-129a7371aba3b1f9`rust_panic at panicking.rs:733:9 [opt]
    frame #1: 0x0000000100aaeb78 sqllogictest-129a7371aba3b1f9`std::panicking::rust_panic_with_hook::h7f09977c23432887 at panicking.rs:703:5 [opt]
    frame #2: 0x0000000100aae790 sqllogictest-129a7371aba3b1f9`std::panicking::begin_panic_handler::_$u7b$$u7b$closure$u7d$$u7d$::h4fa90efaf3565814 at panicking.rs:579:13 [opt]
    frame #3: 0x0000000100aabf5c sqllogictest-129a7371aba3b1f9`std::sys_common::backtrace::__rust_end_short_backtrace::hb60096430afa2290 at backtrace.rs:139:18 [opt]
    frame #4: 0x0000000100aae520 sqllogictest-129a7371aba3b1f9`rust_begin_unwind at panicking.rs:577:5 [opt]
    frame #5: 0x0000000100ae7bd4 sqllogictest-129a7371aba3b1f9`core::panicking::panic_fmt::h2abc5fee73b14749 at panicking.rs:110:14 [opt]
    frame #6: 0x0000000100ae7b58 sqllogictest-129a7371aba3b1f9`core::panicking::panic::hedadeb8ebf54333f at panicking.rs:48:5 [opt]
    frame #7: 0x0000000100546cdc sqllogictest-129a7371aba3b1f9`core::option::Option$LT$T$GT$::unwrap::h4417ac906dd011ba(self=Option<risinglight::storage::secondary::block::blob_block_builder::PlainBlobBlockBuilder<risinglight::types::blob::BlobRef>> @ 0x00000001780eea88) at option.rs:752:21
    frame #8: 0x0000000100490054 sqllogictest-129a7371aba3b1f9`risinglight::storage::secondary::column::blob_column_builder::BlobColumnBuilder::finish_builder::h7bb3921b0dad32bb(self=0x00000001780ef8f0) at blob_column_builder.rs:35:23
    frame #9: 0x00000001004903b0 sqllogictest-129a7371aba3b1f9`_$LT$risinglight..storage..secondary..column..blob_column_builder..BlobColumnBuilder$u20$as$u20$risinglight..storage..secondary..column..ColumnBuilder$LT$risinglight..array..utf8_array..BytesArray$LT$risinglight..types..blob..BlobRef$GT$$GT$$GT$::finish::h534e2234101a26ce(self=BlobColumnBuilder @ 0x00000001780ef8f0) at blob_column_builder.rs:71:9
    frame #10: 0x00000001002acfdc sqllogictest-129a7371aba3b1f9`risinglight::storage::secondary::column::column_builder::ColumnBuilderImpl::finish::hf6b1b5103806d506(self=ColumnBuilderImpl @ 0x00000001780efce0) at column_builder.rs:90:36
    frame #11: 0x0000000100221b64 sqllogictest-129a7371aba3b1f9`risinglight::storage::secondary::rowset::rowset_builder::RowsetBuilder::finish_and_flush::_$u7b$$u7b$closure$u7d$$u7d$::hd935983a5c742dc7((null)=ResumeTy @ 0x00000001780f11c0) at rowset_builder.rs:103:33
    frame #12: 0x0000000100343cc8 sqllogictest-129a7371aba3b1f9`_$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$::poll::hea002ec7321fa057(self=Pin<&mut core::future::from_generator::GenFuture<risinglight::storage::secondary::rowset::rowset_builder::{impl#0}::finish_and_flush::{async_fn#0}>> @ 0x00000001780f13b8, cx=0x00000001780f9420) at mod.rs:84:19
    frame #13: 0x00000001000b6080 sqllogictest-129a7371aba3b1f9`risinglight::storage::secondary::compactor::Compactor::compact_table::_$u7b$$u7b$closure$u7d$$u7d$::h5f15f548bcff608c((null)=ResumeTy @ 0x00000001780f5560) at compactor.rs:109:35
    frame #14: 0x00000001000ae660 sqllogictest-129a7371aba3b1f9`_$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$::poll::h15b46372709aa650(self=Pin<&mut core::future::from_generator::GenFuture<risinglight::storage::secondary::compactor::{impl#0}::compact_table::{async_fn#0}>> @ 0x00000001780f58c8, cx=0x00000001780f9420) at mod.rs:84:19
    frame #15: 0x00000001000b77f0 sqllogictest-129a7371aba3b1f9`risinglight::storage::secondary::compactor::Compactor::run::_$u7b$$u7b$closure$u7d$$u7d$::h02fa82435d928225((null)=ResumeTy @ 0x00000001780f7570) at compactor.rs:162:80
    frame #16: 0x00000001000aed28 sqllogictest-129a7371aba3b1f9`_$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$::poll::h38f138f514ed507a(self=Pin<&mut core::future::from_generator::GenFuture<risinglight::storage::secondary::compactor::{impl#0}::run::{async_fn#0}>> @ 0x00000001780f7848, cx=0x00000001780f9420) at mod.rs:84:19
    frame #17: 0x00000001000687e8 sqllogictest-129a7371aba3b1f9`risinglight::storage::secondary::SecondaryStorage::spawn_compactor::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h59999c38a8bb5890((null)=ResumeTy @ 0x00000001780f9198) at mod.rs:119:27
  ...
```

The debugger said it was merging two rowsets into a new one. I tried to figure out why the builder receives no data chunks before it is flushed but haven't got any idea. My assumption is that the `blob.slt` test deletes all rows twice before it's going to executes the error tests (without insertion) after, which gives the compactor time to compact those rowsets with all rows deleted. 

I'm not an expert on the storage so leave this to who really knows.

### PR DESCRIPTION
**Pull request by arkbriar**
Signed-off-by: arkbriar <arkbriar@gmail.com>

Separates the write logic from the rowset builder and also fixes the #493 . 

close https://github.com/risinglightdb/risinglight/issues/110
close #493

I fix it by skipping the entire add rowset procedure. Not sure if it's legal in all circumstances, e.g. compact all rowsets in a table (empty and delete them all. Is it necessary to make sure as least one active rowset in the directory? @skyzh

### CODE REVIEW

**Comment by skyzh** [comment]
> Is it necessary to make sure as least one active rowset in the directory

No need. I'll take a look at this PR next week :)

File: src/storage/secondary/rowset/mem_rowset.rs [hunk 1]
```diff
@@ -119,8 +120,9 @@ impl MemTable for ColumnMemTable {
 
 pub struct SecondaryMemRowset<M: MemTable> {
     mem_table: M,
-    rowset_builder: RowsetBuilder,
     rowset_id: u32,
+    directory: PathBuf,
```
Comment (zzl200012) [inline src/storage/secondary/rowset/mem_rowset.rs:124]: I think it's better to remove `directory` from memory rowset, and pass the directory as a parameter when doing flush.

File: src/storage/secondary/rowset/mem_rowset.rs [hunk 2]
```diff
@@ -134,7 +136,8 @@ impl SecondaryMemRowset<BTreeMapMemTable> {
         let chunk = self.mem_table.flush()?;
         let mut builder = self.rowset_builder;
         builder.append(chunk);
-        builder.finish_and_flush().await?;
+        let writer = RowsetWriter::new(self.directory);
```
Comment (zzl200012) [inline src/storage/secondary/rowset/mem_rowset.rs:139]: Pass `directory` as a parameter here? So that some upper logic should be adjusted as well.

**Comment by zzl200012** [comment]
Others LGTM! Let @skyzh have a more detailed review then.

File: src/storage/secondary/rowset/mem_rowset.rs (hunk 1 above)
Comment (arkbriar) [reply src/storage/secondary/rowset/mem_rowset.rs:124]: That's better! Thanks!

File: src/storage/secondary/rowset/raw.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+use std::sync::Arc;
+
+use crate::catalog::ColumnCatalog;
+
+/// Raw columns.
```
Comment (skyzh) [inline src/storage/secondary/rowset/raw.rs:5]: What's the meaning of `Raw`? Better to name it as `EncodedColumn` or something?

**Review by skyzh — APPROVED** [review]
LGTM, good work!

File: src/storage/secondary/rowset/raw.rs (hunk 3 above)
Comment (arkbriar) [reply src/storage/secondary/rowset/raw.rs:5]: Intended to be "in raw bytes". "Encoded" is better as with different encoders we get different bytes.
I'll fix it.

**Pull request merged**